### DISCUSSION TITLE
How to add entire source folders to an installer?

### DISCUSSION THREAD

**Discussion by crazydef** [wixtoolset/issues#8002] [Questions]
I'm trying to setup Wix to create an installer in a DevOps pipeline.

After other parts of the pipeline has run, I have a couple of folders, `Bin`, and `Data` that I want packaged into the installer.  Collectively they contain a few hundred files, and a dozen sub folders, so adding references manually isn't an option.

After reading a ton of documentation this morning, it seems I should be able jump through flaming hoops while juggling goats and run `heat` to _harvest_ the folders' contents.  I see `heat` can be installed as a Wix extension, but how do I actually run it?  And can I even run it, given it's not [recommended/supported](https://github.com/wixtoolset/issues/issues/6838) in Wix 4.

I don't have Visual Studio projects or a solution, so I was hoping I could just run everything from a `script` section in the pipeline yaml.

So my question is:  How do I add the entire contents of a folder (including any sub folders) to a Wix installer such that it is simply duplicated into the destination folder under `Program Files` on the target machine when the installer is run?

**Comment by robmen** [comment]
For WiX v4, I cover a couple of options in [Deployment Dojo S1:E53](https://robmensching.com/deployment-dojo/episodes/s1/e53/harvesting-simplified-by-the-heatwave-build-tools-for-wix-v4/) (the easy way) and [S1:E52](https://robmensching.com/deployment-dojo/episodes/s1/e52/harvesting-the-hard-way-with-heat-in-wix-v4/) (the hard way). You might find those useful.

**Comment by achainarong** [comment]
Hey Crazydef,

i can provide you with a small example here like this. When you create your Wix Project with Visual Studio you can add those line to harvest the folders. It's just an example you have to check yourselfe it this works out for you. If you want to run it in the pipeline you can just run "dotnet build" and it will generate the msi.

project_name.wixproj:

`Project Sdk="WixToolset.Sdk/4.0.4">  
  <ItemGroup>
    <PackageReference Include="WixToolset.Heat" Version="4.0.4" />
  </ItemGroup>
  <PropertyGroup>
    <DefineConstants>ApplicationFolder=bin</DefineConstants>
  </PropertyGroup>
  <ItemGroup>
    <HarvestDirectory Include="bin">
      <ComponentGroupName>BinComponenets</ComponentGroupName>
      <DirectoryRefId>ApplicationFolder</DirectoryRefId>
      <SuppressRootDirectory>true</SuppressRootDirectory>
      <SuppressRegistry>true</SuppressRegistry>
    </HarvestDirectory>
    <BindPath Include="bin" />
  </ItemGroup>
</Project>`

Package.wxs:

`<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
  <Package Name="BinDemo" Manufacturer="<Your Company>" Version="1.0.0.0" UpgradeCode="<Some GUID>">
    <MajorUpgrade DowngradeErrorMessage="!(loc.DowngradeError)" />
    <Feature Id="Main">
      <ComponentGroupRef Id="BinComponents" />
    </Feature>
  </Package>
</Wix>
`

**Comment by jjolidon** [comment]
Hi @crazydef! In Wix 5 you could use [Files](https://wixtoolset.org/docs/schema/wxs/files/), which will recursively find all the files in a folder and include them to a component.
